test/new-e2e/shared-components/hostname_ec2_test.go

```diff
@@ -0,0 +1,100 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package testinfradefinition
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/DataDog/datadog-agent/test/new-e2e/utils/e2e"
+	"github.com/DataDog/datadog-agent/test/new-e2e/utils/e2e/client"
+	"github.com/DataDog/test-infra-definitions/components/datadog/agentparams"
+	"github.com/stretchr/testify/assert"
+)
+
+type agentSuite struct {
+	e2e.Suite[e2e.AgentEnv]
+}
+
+func TestAgentHostnameEC2Suite(t *testing.T) {
+	e2e.Run(t, &agentSuite{}, e2e.AgentStackDef(nil))
+}
+
+// https://github.com/DataDog/datadog-agent/blob/main/pkg/util/hostname/README.md#the-current-logic
+func (v *agentSuite) TestAgentHostnameDefaultsToResourceId() {
+	v.UpdateEnv(e2e.AgentStackDef(nil, agentparams.WithAgentConfig("")))
+
+	metadata := client.NewEC2Metadata(v.Env().VM)
+	err := v.Env().Agent.WaitForReady()
+	assert.NoError(v.T(), err)
+
+	hostname := v.Env().Agent.Hostname()
+
+	// Default configuration of hostname for EC2 instances is the resource-id
+	resourceId := metadata.Get("instance-id")
+	assert.Contains(v.T(), hostname, resourceId)
+}
+
+func (v *agentSuite) TestAgentConfigHostnameVarOverride() {
+	v.UpdateEnv(e2e.AgentStackDef(nil, agentparams.WithAgentConfig("hostname: hostname.from.var")))
+	err := v.Env().Agent.WaitForReady()
+	assert.NoError(v.T(), err)
+
+	hostname := v.Env().Agent.Hostname()
+	assert.Contains(v.T(), hostname, "hostname.from.var")
+}
+
+func (v *agentSuite) TestAgentConfigHostnameFileOverride() {
+	fileContent := "hostname.from.file"
+	v.Env().VM.Execute(fmt.Sprintf(`echo "%s" | tee /tmp/hostname`, fileContent))
+	v.UpdateEnv(e2e.AgentStackDef(nil, agentparams.WithAgentConfig("hostname_file: /tmp/hostname")))
+
+	err := v.Env().Agent.WaitForReady()
+	assert.NoError(v.T(), err)
+
+	hostname := v.Env().Agent.Hostname()
+	assert.Contains(v.T(), hostname, fileContent)
+}
+
+// hostname_force_config_as_canonical stops throwing a warning but doesn't change behavior
```

Comment:
I tried to make this test but I don't think it is possible with the framework or it requires more knowledge than I have modifying ec2 host

This is because [the check for `isHostnameCanonicalForIntake`](https://github.com/DataDog/datadog-agent/blob/main/pkg/util/hostname/warnings.go#L19C6-L25) to throw the warning (before the variable is checked) will always be true. It checks to see we can fetch metadata from AWS if the hostname prefix is a default

Reply:
I'm not sure to understand why it's not possible. 

My understanding is that setting 'hostname: ip-1-2-3-4' + the VM  being able to fetch the instance ID is enough to trigger the warning (was able to reproduce). Is the second part blocking ?

Let think about it in a follow up, the current PR is mergeable and it'll be really easy to follow up